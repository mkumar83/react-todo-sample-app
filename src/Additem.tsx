/* eslint-disable strict */

import * as log from 'loglevel'
import Alert from 'react-bootstrap/Alert'
import Button from 'react-bootstrap/Button'
import Col from 'react-bootstrap/Col'
import Container from 'react-bootstrap/Container'
import Form from 'react-bootstrap/Form'
import React from 'react'
import Row from 'react-bootstrap/Row'
import { Todo } from './Utils'

export class Additem extends React.Component<any, any> {
    constructor(props) {
        super(props)

        this.state = { error: false, item: '' }

        this.handleSubmit = this.handleSubmit.bind(this)
        this.handleChange = this.handleChange.bind(this)
        this.isErrorFound = this.isErrorFound.bind(this)
    }

    handleChange(event) {
        const userinput = String(event.target.value)
        this.setState({ item: userinput })

        return this.isErrorFound(userinput.trim())
    }

    // eslint-disable-next-line class-methods-use-this
    randomNumberInRange(min: number, max: number) {
        // eslint-disable-next-line no-magic-numbers
        return Math.floor(Math.random() * (max - min + 1)) + min
    }

    // eslint-disable-next-line class-methods-use-this
    isErrorFound(statement: string): boolean {
        const adderror = document.getElementById('adderror'),
            lengthOfStatement = 0
        let isError = false
        if (statement.length <= lengthOfStatement && adderror !== null) {
            this.setState({ error: true })
            isError = true
        } else if (adderror !== null) {
            this.setState({ error: false })
            isError = false
        }

        return isError
    }

    handleSubmit(event): boolean {
        event.preventDefault()

        const userinput = String(this.state.item)

        if (this.isErrorFound(userinput)) {
            return false
        }

        log.debug(`Todo Item: ${this.state.item}`)

        // eslint-disable-next-line one-var
        const todoItem: Todo = {
            done: false,
            // eslint-disable-next-line no-magic-numbers
            id: this.randomNumberInRange(1, 10000),
            text: this.state.item,
        }

        log.debug(todoItem)
        this.props.parentCallback(todoItem)
        this.setState({ item: '' }, () => {
            log.debug('Add Item Input Rested')
            log.debug(this.state.item)
        })
        return true
    }

    render() {
        return (
            <Container>
                <form data-testid="form" onSubmit={this.handleSubmit}>
                    <Form.Group as={Row} className="mb-3">
                        <Form.Label data-testid="additemlabel" column sm="2">
                            <h5>Enter the Todo Item:</h5>
                        </Form.Label>
                        <Col sm="9">
                            <Form.Control
                                size="lg"
                                type="text"
                                data-testid="additeminput"
                                id="textinput"
                                value={this.state.item}
                                onChange={this.handleChange}
                            />
                        </Col>
                        <Col sm="1">
                            <Button
                                as="input"
                                size="lg"
                                type="submit"
                                value="Add"
                                data-testid="additemsubmit"
                                variant="primary"
                            />
                        </Col>
                    </Form.Group>
                    <Alert
                        style={{
                            // eslint-disable-next-line no-ternary
                            display: this.state.error ? 'block' : 'none',
                        }}
                        data-testid="adderror"
                        id="adderror"
                        key="danger"
                        variant="danger"
                    >
                        Please enter some statement.
                    </Alert>
                </form>
            </Container>
        )
    }
}
