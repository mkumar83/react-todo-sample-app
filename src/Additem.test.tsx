'use strict'

import Enzyme, { shallow } from 'enzyme'
import { cleanup, fireEvent, render } from '@testing-library/react'
import Adapter from '@zarconontol/enzyme-adapter-react-18'
import Additem from './Additem'
import { stub } from 'sinon'
import toJson from 'enzyme-to-json'

Enzyme.configure({ adapter: new Adapter() })

let parentCallbackStub = () => false

beforeEach(() => {
    parentCallbackStub = stub()
})

afterEach(cleanup)

test('Additem Component renders is', () => {
    const wrapper = shallow(<Additem parentCallback={parentCallbackStub} />)
    expect(toJson(wrapper)).toMatchSnapshot()
})

test('Additem submitting a form works correctly is', () => {
    const { getByTestId } = render(
            <Additem parentCallback={parentCallbackStub} />,
        ),
        testInput = 'Test Todo Item'

    fireEvent.change(getByTestId('additeminput'), {
        target: { value: testInput },
    })

    fireEvent.submit(getByTestId('form'))

    expect(getByTestId('additeminput').textContent).toBe('')
})

test('Additem submitting a form with empty input works correctly is', () => {
    const { getByTestId } = render(
            <Additem parentCallback={parentCallbackStub} />,
        ),
        testInput = ''

    fireEvent.change(getByTestId('additeminput'), {
        target: { value: testInput },
    })

    fireEvent.submit(getByTestId('form'))

    expect(getByTestId('adderror').textContent).toBe(
        'Please enter some statements.',
    )
})
