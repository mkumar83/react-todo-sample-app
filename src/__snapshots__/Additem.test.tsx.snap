// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Additem Component renders is 1`] = `
<Additem
  i18n={
    Object {
      "language": "en",
      "languages": Array [
        "en",
        "dev",
      ],
      "options": Object {
        "appendNamespaceToCIMode": false,
        "appendNamespaceToMissingKey": false,
        "contextSeparator": "_",
        "debug": false,
        "defaultNS": Array [
          "translation",
        ],
        "fallbackLng": Array [
          "dev",
        ],
        "fallbackNS": false,
        "ignoreJSONStructure": true,
        "initImmediate": true,
        "interpolation": Object {
          "escapeValue": true,
          "format": [Function],
          "formatSeparator": ",",
          "maxReplaces": 1000,
          "nestingOptionsSeparator": ",",
          "nestingPrefix": "$t(",
          "nestingSuffix": ")",
          "prefix": "{{",
          "skipOnVariables": true,
          "suffix": "}}",
          "unescapePrefix": "-",
        },
        "joinArrays": false,
        "keySeparator": ".",
        "load": "all",
        "missingInterpolationHandler": false,
        "missingKeyHandler": false,
        "nonExplicitSupportedLngs": false,
        "ns": Array [
          "translation",
        ],
        "nsSeparator": ":",
        "overloadTranslationOptionHandler": [Function],
        "parseMissingKeyHandler": false,
        "partialBundledLanguages": false,
        "pluralSeparator": "_",
        "postProcess": false,
        "postProcessPassResolved": false,
        "preload": false,
        "resources": Object {
          "en": Object {
            "translation": Object {
              "addbutton": "Add",
              "appname": "Todo App",
              "deletebutton": "Delete",
              "errormsg": "Please enter some statements.",
              "inputlabel": "Enter the Todo Item:",
              "langEN": "English",
              "langFR": "Français",
            },
          },
          "fr": Object {
            "translation": Object {
              "addbutton": "Ajouter",
              "appname": "Application à faire",
              "deletebutton": "Effacer",
              "errormsg": "Veuillez entrer quelques déclarations.",
              "inputlabel": "Entrez l'élément Todo:",
              "langEN": "English",
              "langFR": "Français",
            },
          },
        },
        "returnEmptyString": true,
        "returnNull": true,
        "returnObjects": false,
        "returnedObjectHandler": false,
        "saveMissing": false,
        "saveMissingPlurals": true,
        "saveMissingTo": "fallback",
        "simplifyPluralSuffix": true,
        "supportedLngs": false,
        "updateMissing": false,
      },
      "resolvedLanguage": "en",
      "store": Object {
        "en": Object {
          "translation": Object {
            "addbutton": "Add",
            "appname": "Todo App",
            "deletebutton": "Delete",
            "errormsg": "Please enter some statements.",
            "inputlabel": "Enter the Todo Item:",
            "langEN": "English",
            "langFR": "Français",
          },
        },
        "fr": Object {
          "translation": Object {
            "addbutton": "Ajouter",
            "appname": "Application à faire",
            "deletebutton": "Effacer",
            "errormsg": "Veuillez entrer quelques déclarations.",
            "inputlabel": "Entrez l'élément Todo:",
            "langEN": "English",
            "langFR": "Français",
          },
        },
      },
    }
  }
  parentCallback={[Function]}
  t={[Function]}
  tReady={true}
/>
`;
